opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -12,7 +12,214 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
+from collections import OrderedDict
+from collections import deque
+from collections import namedtuple
+import threading
+import typing
+
 from opentelemetry import trace as trace_api
+from opentelemetry.sdk import util
+
+try:
+    from collections.abc import MutableMapping
+    from collections.abc import Sequence
+except ImportError:
+    from collections import MutableMapping
+    from collections import Sequence
+
+
+MAX_NUM_ATTRIBUTES = 32
+MAX_NUM_ANNOTATIONS = 32
+MAX_NUM_EVENTS = 128
+MAX_NUM_LINKS = 32
+
+AttributeValue = typing.Union[str, bool, float]
+
+
+class BoundedList(Sequence):
+    """An append only list with a fixed max size."""
+    def __init__(self, maxlen):
+        self.dropped = 0
+        self._dq = deque(maxlen=maxlen)
+        self._lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({}, maxlen={})"
+                .format(
+                    type(self).__name__,
+                    list(self._dq),
+                    self._dq.maxlen
+                ))
+
+    def __getitem__(self, index):
+        return self._dq[index]
+
+    def __len__(self):
+        return len(self._dq)
+
+    def __iter__(self):
+        return iter(self._dq)
+
+    def append(self, item):
+        with self._lock:
+            if len(self._dq) == self._dq.maxlen:
+                self.dropped += 1
+            self._dq.append(item)
+
+    def extend(self, seq):
+        with self._lock:
+            to_drop = len(seq) + len(self._dq) - self._dq.maxlen
+            if to_drop > 0:
+                self.dropped += to_drop
+            self._dq.extend(seq)
+
+    @classmethod
+    def from_seq(cls, maxlen, seq):
+        seq = tuple(seq)
+        if len(seq) > maxlen:
+            raise ValueError
+        bounded_list = cls(maxlen)
+        bounded_list._dq = deque(seq, maxlen=maxlen)
+        return bounded_list
+
+
+class BoundedDict(MutableMapping):
+    """A dict with a fixed max capacity."""
+    def __init__(self, maxlen):
+        self.maxlen = maxlen
```

Comment:
Need to validate the input `isinstance(maxlen, int) and maxlen >= 0` (`maxlen == 0` is allowed for consistency with `deque(maxlen=0)`).